File: matlab/optiprofiler/tests/action_tests/random_test.m
function random_test(benchmark_id)
%RANDOM_TEST tests OptiProfiler with random options.

    % Use the current wall-clock time to generate a random seed.
    time = datetime;
    seed = 100*mod(year(time), 100) + month(time) + week(time) + day(time);
    fprintf("Seed: %d\n\n", seed);
    rand_stream = RandStream("mt19937ar", "Seed", seed);

    % Use the random seed to generate random options.
    options = struct();

    n_jobs_choices = (1:5);
    options.n_jobs = n_jobs_choices(rand_stream.randi(length(n_jobs_choices), 1, 1));

    options.seed = seed;

    options.benchmark_id = benchmark_id;

    solvers_choices = {@fmincon_test1, @fmincon_test2, @fmincon_test3};
    solver_names_choices = {'sqp', 'interior-point', 'active-set'};
    solver_2_or_3 = rand_stream.randi([2, 3], 1, 1);
    if solver_2_or_3 == 2
        idx = rand_stream.randperm(2, 2);
        solvers = solvers_choices(idx);
        options.solver_names = solver_names_choices(idx);
    else
        solvers = solvers_choices;
        options.solver_names = solver_names_choices;
    end

    error_bar_choices = {'minmax', 'meanstd'};
    options.errorbar_type = error_bar_choices{rand_stream.randi(length(error_bar_choices), 1, 1)};

    max_tol_order_choices = (1:16);
    options.max_tol_order = max_tol_order_choices(rand_stream.randi(length(max_tol_order_choices), 1, 1));

    options.max_eval_factor = rand_stream.rand(1, 1) * 1000;

    options.project_x0 = (rand_stream.rand(1, 1) < 0.5);

    options.run_plain = (rand_stream.rand(1, 1) < 0.5);

    options.score_only = (rand_stream.rand(1, 1) < 0.5);

    options.summarize_performance_profiles = (rand_stream.rand(1, 1) < 0.5);

    options.summarize_data_profiles = (rand_stream.rand(1, 1) < 0.5);

    if length(solvers) == 2
        options.summarize_log_ratio_profiles = (rand_stream.rand(1, 1) < 0.5);
    else
        options.summarize_log_ratio_profiles = false;
    end

    options.summarize_output_based_profiles = (rand_stream.rand(1, 1) < 0.5);

    options.solver_verbose = rand_stream.randi([0, 2], 1, 1);

    options.semilogx = (rand_stream.rand(1, 1) < 0.5);

    options.normalized_scores = (rand_stream.rand(1, 1) < 0.5);

    line_colors_choices = {'r', 'g', 'b', 'c', 'm', 'y', 'k'};
    options.line_colors = line_colors_choices(rand_stream.randperm(length(line_colors_choices), length(solvers)));

    line_styles_choices = {...
    '-', '-o', '-+', '-*', '-.', '-x', '-s', '-d', '-^', '-v', '->', '-<', '-p', '-h', 'o-', '+-', '*-', '.-', 'x-', 's-', 'd-', '^-', 'v-', '>-', '<-', 'p-', 'h-',...
    '-.', '-.o', '-.+', '-.*', '-..', '-.x', '-.s', '-.d', '-.^', '-.v', '-.>', '-.<', '-.p', '-.h', 'o-.', '+-.', '*-.', '.-.', 'x-.', 's-.', 'd-.', '^-.', 'v-.', '>-.', '<-.', 'p-.', 'h-.',...
    ':', ':o', ':+', ':*', ':.', ':x', ':s', ':d', ':^', ':v', ':>', ':<', ':p', ':h', 'o:', '+:', '*:', '.:', 'x:', 's:', 'd:', '^:', 'v:', '>:', '<:', 'p:', 'h:',...
    '--', '--o', '--+', '--*', '--.', '--x', '--s', '--d', '--^', '--v', '-->', '--<', '--p', '--h', 'o--', '+--', '*--', '.--', 'x--', 's--', 'd--', '^--', 'v--', '>--', '<--', 'p--', 'h--'...
    };
    options.line_styles = line_styles_choices(rand_stream.randperm(length(line_styles_choices), length(solvers)));

    options.line_widths = rand_stream.rand(1, length(solvers)) * 2 + 0.5;

    bar_colors_choices = {'r', 'g', 'b', 'c', 'm', 'y', 'k'};
    options.bar_colors = bar_colors_choices(rand_stream.randperm(length(bar_colors_choices), length(solvers)));

    feature_name_choices = {'plain', 'perturbed_x0', 'noisy', 'truncated', 'permuted', 'linearly_transformed', 'random_nan', 'unrelaxable_constraints', 'nonquantifiable_constraints', 'quantized'};
    options.feature_name = feature_name_choices{rand_stream.randi(length(feature_name_choices), 1, 1)};

    switch options.feature_name
        case 'plain'
            options.n_runs = 1;
        case 'perturbed_x0'
            options.n_runs = rand_stream.randi([1, 3], 1, 1);
            options.perturbation_level = rand_stream.rand(1, 1);
            distribution_choices = {'gaussian', 'spherical'};
            options.distribution = distribution_choices{rand_stream.randi(length(distribution_choices), 1, 1)};
        case 'noisy'
            options.n_runs = rand_stream.randi([1, 3], 1, 1);
            options.noise_level = rand_stream.rand(1, 1);
            distribution_choices = {'gaussian', 'uniform'};
            options.distribution = distribution_choices{rand_stream.randi(length(distribution_choices), 1, 1)};
            noise_type_choices = {'absolute', 'relative', 'mixed'};
            options.noise_type = noise_type_choices{rand_stream.randi(length(noise_type_choices), 1, 1)};
        case 'truncated'
            options.n_runs = rand_stream.randi([1, 3], 1, 1);
            options.significant_digits = rand_stream.randi([1, 10], 1, 1);
            options.perturbed_trailing_digits = (rand_stream.rand(1, 1) < 0.5);
        case 'linearly_transformed'
            options.n_runs = rand_stream.randi([1, 3], 1, 1);
            options.rotated = (rand_stream.rand(1, 1) < 0.5);
            options.condition_factor = rand_stream.rand(1, 1) * 100;
        case 'random_nan'
            options.n_runs = rand_stream.randi([1, 3], 1, 1);
            options.nan_rate = rand_stream.rand(1, 1) * 0.5;
        case 'unrelaxable_constraints'
            options.n_runs = 1;
            options.unrelaxable_bounds = (rand_stream.rand(1, 1) < 0.5);
            options.unrelaxable_linear_constraints = (rand_stream.rand(1, 1) < 0.5);
            options.unrelaxable_nonlinear_constraints = (rand_stream.rand(1, 1) < 0.5);
        case 'nonquantifiable_constraints'
            options.n_runs = 1;
        case 'quantized'
            options.n_runs = 1;
            options.mesh_size = rand_stream.rand(1, 1);
            mesh_type_choices = {'absolute', 'relative'};
            options.mesh_type = mesh_type_choices{rand_stream.randi(length(mesh_type_choices), 1, 1)};
            options.ground_truth = (rand_stream.rand(1, 1) < 0.5);
    end

    if isunix() && ~ismac()
        options.plibs = {'s2mpj', 'matcutest'};
    else
        options.plibs = {'s2mpj'};
    end

    ptype_choices = {'u', 'b', 'l', 'n', 'ub', 'ul', 'un', 'bl', 'bn', 'ln', 'ubl', 'ubn', 'uln', 'bln', 'ubln'};
    options.ptype = ptype_choices{rand_stream.randi(length(ptype_choices), 1, 1)};

    options.mindim = rand_stream.randi([1, 3], 1, 1);
    options.maxdim = options.mindim + rand_stream.randi([1, 3], 1, 1);

    options.minb = 0;
    options.maxb = rand_stream.randi([20, 40], 1, 1);

    options.minlcon = 0;
    options.maxlcon = rand_stream.randi([10, 20], 1, 1);

    options.minnlcon = 0;
    options.maxnlcon = rand_stream.randi([10, 20], 1, 1);

    options.mincon = 0;
    options.maxcon = rand_stream.randi([20, 40], 1, 1);

    if rand_stream.rand(1, 1) < 0.5
        options.xlabel_data_profile = '';
    end

    if rand_stream.rand(1, 1) < 0.5
        options.ylabel_data_profile = '';
    end

    if rand_stream.rand(1, 1) < 0.5
        options.xlabel_performance_profile = '';
    end

    if rand_stream.rand(1, 1) < 0.5
        options.ylabel_performance_profile = '';
    end

    if rand_stream.rand(1, 1) < 0.5
        options.xlabel_log_ratio_profile = '';
    end

    if rand_stream.rand(1, 1) < 0.5
        options.ylabel_log_ratio_profile = '';
    end

    % Run the stress test.
    benchmark(solvers, options)
end

function x = fmincon_test1(varargin)
    options = optimoptions('fmincon', 'SpecifyObjectiveGradient', false, 'Algorithm', 'sqp');
    if nargin == 2
        fun = varargin{1};
        x0 = varargin{2};
        x = fmincon(fun, x0, [], [], [], [], [], [], [], options);
    elseif nargin == 4
        fun = varargin{1};
        x0 = varargin{2};
        xl = varargin{3};
        xu = varargin{4};
        x = fmincon(fun, x0, [], [], [], [], xl, xu, [], options);
    elseif nargin == 8
        fun = varargin{1};
        x0 = varargin{2};
        xl = varargin{3};
        xu = varargin{4};
        aub = varargin{5};
        bub = varargin{6};
        aeq = varargin{7};
        beq = varargin{8};
        x = fmincon(fun, x0, aub, bub, aeq, beq, xl, xu, [], options);
    elseif nargin == 10
        fun = varargin{1};
        x0 = varargin{2};
        xl = varargin{3};
        xu = varargin{4};
        aub = varargin{5};
        bub = varargin{6};
        aeq = varargin{7};
        beq = varargin{8};
        cub = varargin{9};
        ceq = varargin{10};
        nonlcon = @(x) deal(cub(x), ceq(x));
        x = fmincon(fun, x0, aub, bub, aeq, beq, xl, xu, nonlcon, options);
    end
end

function x = fmincon_test2(varargin)

    options = optimoptions('fmincon', 'SpecifyObjectiveGradient', false, 'Algorithm', 'interior-point');
    if nargin == 2
        fun = varargin{1};
        x0 = varargin{2};
        x = fmincon(fun, x0, [], [], [], [], [], [], [], options);
    elseif nargin == 4
        fun = varargin{1};
        x0 = varargin{2};
        xl = varargin{3};
        xu = varargin{4};
        x = fmincon(fun, x0, [], [], [], [], xl, xu, [], options);
    elseif nargin == 8
        fun = varargin{1};
        x0 = varargin{2};
        xl = varargin{3};
        xu = varargin{4};
        aub = varargin{5};
        bub = varargin{6};
        aeq = varargin{7};
        beq = varargin{8};
        x = fmincon(fun, x0, aub, bub, aeq, beq, xl, xu, [], options);
    elseif nargin == 10
        fun = varargin{1};
        x0 = varargin{2};
        xl = varargin{3};
        xu = varargin{4};
        aub = varargin{5};
        bub = varargin{6};
        aeq = varargin{7};
        beq = varargin{8};
        cub = varargin{9};
        ceq = varargin{10};
        nonlcon = @(x) deal(cub(x), ceq(x));
        x = fmincon(fun, x0, aub, bub, aeq, beq, xl, xu, nonlcon, options);
    end
end

function x = fmincon_test3(varargin)
    options = optimoptions('fmincon', 'SpecifyObjectiveGradient', false, 'Algorithm', 'active-set');
    if nargin == 2
        fun = varargin{1};
        x0 = varargin{2};
        x = fmincon(fun, x0, [], [], [], [], [], [], [], options);
    elseif nargin == 4
        fun = varargin{1};
        x0 = varargin{2};
        xl = varargin{3};
        xu = varargin{4};
        x = fmincon(fun, x0, [], [], [], [], xl, xu, [], options);
    elseif nargin == 8
        fun = varargin{1};
        x0 = varargin{2};
        xl = varargin{3};
        xu = varargin{4};
        aub = varargin{5};
        bub = varargin{6};
        aeq = varargin{7};
        beq = varargin{8};
        x = fmincon(fun, x0, aub, bub, aeq, beq, xl, xu, [], options);
    elseif nargin == 10
        fun = varargin{1};
        x0 = varargin{2};
        xl = varargin{3};
        xu = varargin{4};
        aub = varargin{5};
        bub = varargin{6};
        aeq = varargin{7};
        beq = varargin{8};
        cub = varargin{9};
        ceq = varargin{10};
        nonlcon = @(x) deal(cub(x), ceq(x));
        x = fmincon(fun, x0, aub, bub, aeq, beq, xl, xu, nonlcon, options);
    end
end

File: python/optiprofiler/action_tests/random_test.py
#!/usr/bin/env python3
"""Random test for OptiProfiler.

Tests OptiProfiler with randomly generated options.
This test is run by GitHub Actions workflows, not by pytest.
"""
import sys
import os
import platform
import numpy as np
from datetime import datetime

# Add parent directory to path for imports
sys.path.insert(0, os.path.dirname(os.path.dirname(os.path.dirname(os.path.dirname(os.path.abspath(__file__))))))

from optiprofiler import benchmark
from optiprofiler.action_tests.solvers import SOLVERS, SOLVER_NAMES


def get_plibs():
    """Get problem libraries based on OS."""
    if sys.platform.startswith('linux') or sys.platform == 'darwin':
        return ['s2mpj', 'pycutest']
    else:
        return ['s2mpj']


def get_benchmark_id():
    """Generate a benchmark ID based on OS and Python version."""
    os_name = platform.system().lower()
    py_version = f"py{sys.version_info.major}{sys.version_info.minor}"
    return f"os_{os_name}_{py_version}"


def random_test(benchmark_id=None):
    """Run random test with randomly generated options."""
    if benchmark_id is None:
        benchmark_id = get_benchmark_id()
    
    # Generate seed based on current date
    now = datetime.now()
    seed = 100 * (now.year % 100) + now.isocalendar()[1] + now.day
    print(f"Seed: {seed}\n")
    
    rng = np.random.default_rng(seed)
    
    # Build random options
    options = {
        'n_jobs': rng.integers(1, 6),
        'seed': seed,
        'benchmark_id': benchmark_id,
    }
    
    # Random solver selection (2 or 3 solvers)
    num_solvers = rng.integers(2, 4)
    solver_indices = rng.permutation(len(SOLVERS))[:num_solvers].tolist()
    solvers = [SOLVERS[i] for i in solver_indices]
    options['solver_names'] = [SOLVER_NAMES[i] for i in solver_indices]
    
    # Random profile options
    options['errorbar_type'] = rng.choice(['minmax', 'meanstd'])
    options['hist_aggregation'] = rng.choice(['min', 'mean', 'max'])
    options['max_tol_order'] = rng.integers(1, 17)
    options['max_eval_factor'] = rng.random() * 30
    options['project_x0'] = rng.random() < 0.5
    options['run_plain'] = rng.random() < 0.5
    options['score_only'] = rng.random() < 0.5
    options['summarize_performance_profiles'] = rng.random() < 0.5
    options['summarize_data_profiles'] = rng.random() < 0.5
    options['summarize_log_ratio_profiles'] = len(solvers) == 2 and rng.random() < 0.5
    options['summarize_output_based_profiles'] = rng.random() < 0.5
    options['solver_verbose'] = rng.integers(0, 3)
    options['semilogx'] = rng.random() < 0.5
    options['normalized_scores'] = rng.random() < 0.5
    
    # Random line styles and colors
    colors = ['r', 'g', 'b', 'c', 'm', 'y', 'k']
    options['line_colors'] = list(rng.choice(colors, len(solvers), replace=False))
    options['bar_colors'] = list(rng.choice(colors, len(solvers), replace=False))
    
    styles = ['-', '--', '-.', ':']
    options['line_styles'] = list(rng.choice(styles, len(solvers), replace=True))
    options['line_widths'] = (rng.random(len(solvers)) * 2 + 0.5).tolist()
    
    # Random hist plot mode
    hist_choices = ['none', 'sequential', 'parallel']
    options['draw_hist_plots'] = rng.choice(hist_choices)
    
    # Random feature
    feature_choices = [
        'plain', 'perturbed_x0', 'noisy', 'truncated', 'permuted',
        'linearly_transformed', 'random_nan', 'unrelaxable_constraints',
        'nonquantifiable_constraints', 'quantized'
    ]
    feature_name = rng.choice(feature_choices)
    options['feature_name'] = feature_name
    
    # Feature-specific options
    if feature_name == 'plain':
        options['n_runs'] = 1
    elif feature_name == 'perturbed_x0':
        options['n_runs'] = rng.integers(1, 4)
        options['perturbation_level'] = rng.random()
        options['distribution'] = rng.choice(['gaussian', 'spherical'])
    elif feature_name == 'noisy':
        options['n_runs'] = rng.integers(1, 4)
        options['noise_level'] = rng.random()
        options['distribution'] = rng.choice(['gaussian', 'uniform'])
        options['noise_type'] = rng.choice(['absolute', 'relative', 'mixed'])
    elif feature_name == 'truncated':
        options['n_runs'] = rng.integers(1, 4)
        options['significant_digits'] = rng.integers(1, 11)
        options['perturbed_trailing_digits'] = rng.random() < 0.5
    elif feature_name == 'linearly_transformed':
        options['n_runs'] = rng.integers(1, 4)
        options['rotated'] = rng.random() < 0.5
        options['condition_factor'] = rng.random() * 100
    elif feature_name == 'random_nan':
        options['n_runs'] = rng.integers(1, 4)
        options['nan_rate'] = rng.random() * 0.5
    elif feature_name == 'unrelaxable_constraints':
        options['n_runs'] = 1
        options['unrelaxable_bounds'] = rng.random() < 0.5
        options['unrelaxable_linear_constraints'] = rng.random() < 0.5
        options['unrelaxable_nonlinear_constraints'] = rng.random() < 0.5
    elif feature_name == 'nonquantifiable_constraints':
        options['n_runs'] = 1
    elif feature_name == 'quantized':
        options['n_runs'] = 1
        options['mesh_size'] = rng.random()
        options['mesh_type'] = rng.choice(['absolute', 'relative'])
        options['ground_truth'] = rng.random() < 0.5
    
    # Problem library options
    options['plibs'] = get_plibs()
    
    # Random problem type
    ptype_choices = ['u', 'b', 'l', 'n', 'ub', 'ul', 'un', 'bl', 'bn', 'ln', 
                     'ubl', 'ubn', 'uln', 'bln', 'ubln']
    options['ptype'] = rng.choice(ptype_choices)
    
    # Random dimension constraints
    options['mindim'] = rng.integers(1, 4)
    options['maxdim'] = options['mindim'] + rng.integers(1, 4)
    
    # Random constraint counts
    options['minb'] = 0
    options['maxb'] = rng.integers(20, 41)
    options['minlcon'] = 0
    options['maxlcon'] = rng.integers(10, 21)
    options['minnlcon'] = 0
    options['maxnlcon'] = rng.integers(10, 21)
    options['mincon'] = 0
    options['maxcon'] = rng.integers(20, 41)
    
    # Random label options
    if rng.random() < 0.5:
        options['xlabel_data_profile'] = ''
    if rng.random() < 0.5:
        options['ylabel_data_profile'] = ''
    if rng.random() < 0.5:
        options['xlabel_performance_profile'] = ''
    if rng.random() < 0.5:
        options['ylabel_performance_profile'] = ''
    if rng.random() < 0.5:
        options['xlabel_log_ratio_profile'] = ''
    if rng.random() < 0.5:
        options['ylabel_log_ratio_profile'] = ''
    
    # Print options
    print("Running random test with options:")
    for key, value in options.items():
        print(f"  {key}: {value}")
    print()
    
    # Run benchmark
    benchmark(solvers, **options)


if __name__ == '__main__':
    benchmark_id = sys.argv[1] if len(sys.argv) > 1 else None
    random_test(benchmark_id)
